API Refactor Form request handling into FormRequestHandler

tractorcow opened the pull request:
Fixes https://github.com/silverstripe/silverstripe-framework/issues/5639
Fixes https://github.com/silverstripe/silverstripe-framework/issues/6362

Requires https://github.com/silverstripe/silverstripe-cms/pull/1758

dhensby commented:
There are some CS failures

dhensby commented on src/Control/RequestHandler.php:
```diff
@@ -546,4 +547,114 @@ public function setRequest($request)
+        if (!$pageURL) {
+            return null;
+        }
+        if (Director::is_site_url($pageURL)) {
```
Why should the `getReferer` be opinionated about whether the url is a site url or not?

tractorcow replied:
Well, I can call it `getInternalReferer`, but the intent of this method is to get only a referer that's safe to redirect to.

I could add a flag to it to allow external urls to be allowed?

tractorcow replied:
```
/**
 * Get referrer to redirect back to and safely validates it
 */
```

tractorcow replied:
I'll split it into getReferer() (raw) and getReturnReferer() (safely validates)

dhensby replied:
Maybe `getSafeReferer` is all we need as the code to get the referer header is a one liner anyway?

dhensby commented on src/Security/Security.php:
```diff
@@ -2,17 +2,15 @@
 
 namespace SilverStripe\Security;
 
+use Page;
```
`Page` is never used in this class (apart from for a phpdoc type hint)

tractorcow replied:
I'll change this reference to a class reference.
`private static $page_class = Page::class;`

tractorcow commented:
Fixed up linting issues, split getReferer() into getReferer() and getReturnReferer(), and replaced a string literal with class constant in Security.php.

dhensby commented:
The CMS PR needs fixing before we can merge

tractorcow commented:
Fixed @dhensby
